core/src/test/java/org/apache/iceberg/TestBaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,122 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestBaseIncrementalAppendScan extends ScanTestBase<IncrementalAppendScan> {
+  public TestBaseIncrementalAppendScan(int formatVersion) {
+    super(formatVersion);
+  }
+
+  @Override
+  protected IncrementalAppendScan newScan() {
+    return table.newIncrementalAppendScan();
+  }
+
+  @Test
+  public void testFromSnapshotInclusive() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotAId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    long snapshotBId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    long snapshotCId = table.currentSnapshot().snapshotId();
+
+    IncrementalAppendScan scan = newScan()
+        .fromSnapshotInclusive(snapshotAId);
+    Assert.assertEquals(3, Iterables.size(scan.planFiles()));
+  }
+
+  @Test
+  public void testFromSnapshotExclusiveAndToSnapshot() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotAId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    long snapshotBId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    long snapshotCId = table.currentSnapshot().snapshotId();
+
+    IncrementalAppendScan scan = newScan()
+        .fromSnapshotExclusive(snapshotAId)
+        .toSnapshot(snapshotBId);
+    Assert.assertEquals(1, Iterables.size(scan.planFiles()));
+  }
+
+  @Test
+  public void testToSnapshot() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotAId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    long snapshotBId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    long snapshotCId = table.currentSnapshot().snapshotId();
+
+    IncrementalAppendScan scan = newScan()
+        .toSnapshot(snapshotBId);
+    Assert.assertEquals(2, Iterables.size(scan.planFiles()));
+  }
+
+  @Test
+  public void testMultipleRootSnapshots() throws Exception {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotAId = table.currentSnapshot().snapshotId();
+    long expireTimestampSnapshotA = TestHelpers.waitUntilAfter(table.currentSnapshot().timestampMillis());
+
+    // append file B in a staged branch
+    table.newFastAppend().appendFile(FILE_B).stageOnly().commit();
```

Comment:
You can also call the operation's `apply()` method to get the snapshot ID early.

```java
AppendFiles append = table.newFastAppend()...;
Snapshot snapshotWithB = append.apply();
append.commit();
```